### PR TITLE
feat: Add ListDeals and ListDataCertificates for Genesis State

### PR DESCRIPTION
**Pull request by inchori**
#### What I done ####
- Add ```ListDeals``` and ```ListDataCertificates``` for genesis state.
- Test Code for ```ListDeals``` and ```ListDataCertificates``` in ```keeper/deal_test.go```.
- Add ```ExportGenesis``` in ```x/market/genesis.go```.
- Test Code for ```ExportGenesis``` in ```x/market/genesis_test.go```.

#### TODO ####
- ```GetDataCertificateListByDealId``` in next PR.

### CODE REVIEW

File: x/market/genesis.go [hunk 1]
```diff
@@ -22,14 +22,34 @@ func InitGenesis(ctx sdk.Context, k keeper.Keeper, genState types.GenesisState)
...
 func ExportGenesis(ctx sdk.Context, k keeper.Keeper) *types.GenesisState {
-	//genesis := types.DefaultGenesis()
+	deals, err := k.ListDeals(ctx)
+	if err != nil {
+		panic(err)
+	}
 
 	dealsMap := make(map[uint64]*types.Deal)
+	for _, deal := range deals {
+		deal, err := k.GetDeal(ctx, deal.DealId)
```
[the remark this answers is not in the document]
Comment (inchori) [reply x/market/genesis.go:32]: Ah the ```ListDeals``` returns the slice of ```types.Deal```. I committed twice, so you can refer this [e3096a0](https://github.com/medibloc/panacea-core/pull/268/commits/e3096a02c0ac8ea5aca97a61897f3a4e3e4f5fc2) commit.

File: x/market/genesis_test.go [hunk 2]
```diff
@@ -66,6 +70,47 @@ func (suite *genesisTestSuite) TestMarketInitGenesis() {
 	suite.Require().Equal(newDataCert.UnsignedCert.GetRequesterAddress(), dataCertificateStored.UnsignedCert.GetRequesterAddress())
 }
 
+func (suite *genesisTestSuite) TestMarketExportGenesis() {
+	newDeal := makeTestDeal()
+	newDataCert := makeTestDataCert()
+
+	dataCertificateKey := types.GetKeyPrefixDataCertificate(newDataCert.UnsignedCert.DealId, newDataCert.UnsignedCert.DataHash)
+	stringDataCertificateKey := string(dataCertificateKey[:])
```
[the remark this answers is not in the document]
Comment (inchori) [reply x/market/genesis_test.go:78]: I thinks it's not necessary to use ```[:]```. I'll remove it.

File: x/market/keeper/deal_test.go [hunk 3]
```diff
@@ -89,6 +89,48 @@ func (suite *dealTestSuite) TestGetDeal() {
 	suite.Require().Equal(deal.GetStatus(), testDeal.GetStatus())
 }
 
+func (suite *dealTestSuite) TestListDeals() {
```
[the remark this answers is not in the document]
Comment (inchori) [reply x/market/keeper/deal_test.go:92]: Yes, it's correct because I set a deal with dealId of 1 in ```BeforeTest```.

File: x/market/genesis.go [hunk 4]
```diff
@@ -22,14 +22,34 @@ func InitGenesis(ctx sdk.Context, k keeper.Keeper, genState types.GenesisState)
...
+		if err != nil {
+			panic(err)
+		}
```
Comment (gyuguen) [inline x/market/genesis.go:35]:
```suggestion

```
How do you feel about being able to delete this?

File: x/market/genesis.go [hunk 5]
```diff
@@ -22,14 +22,34 @@ func InitGenesis(ctx sdk.Context, k keeper.Keeper, genState types.GenesisState)
...
-	//TODO: Implement GetDealsList and GetDataCertificateByDealId Mapping to dealsMap and dataCertificateMap.
-	//for _, deal := k.GetDeals(ctx) {
-	//}
+	for _, dataCertificate := range dataCertificates {
+		certificate, err := k.GetDataCertificate(ctx, dataCertificate)
+		if err != nil {
+			panic(err)
+		}
+		dataKey := types.CombineKeys(sdk.Uint64ToBigEndian(certificate.UnsignedCert.DealId), certificate.UnsignedCert.DataHash)
+		dataCertificateMap[string(dataKey)] = &certificate
```
Comment (gyuguen) [inline x/market/genesis.go:51]:
```suggestion
		dataKey := types.CombineKeys(sdk.Uint64ToBigEndian(dataCertificate.UnsignedCert.DealId), dataCertificate.UnsignedCert.DataHash)
		dataCertificateMap[string(dataKey)] = &dataCertificate
```

I think you can write it like this.

**Review by gyuguen — APPROVED** [review]
Please check only a few reviews.

**Pull request merged**